diff --git a/CMakeLists.txt b/CMakeLists.txt
index 09f0b84..e5f1c13 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -3,8 +3,9 @@
 #       Unix: export LUA_DIR=/home/user/pkg
 #       Windows: set LUA_DIR=c:\lua51
 
+cmake_minimum_required(VERSION 3.5)
 project(lua-cjson C)
-cmake_minimum_required(VERSION 2.6)
+set(PROJECT_VERSION 2.1.0)
 
 option(USE_INTERNAL_FPCONV "Use internal strtod() / g_fmt() code for performance")
 option(MULTIPLE_THREADS "Support multi-threaded apps with internal fpconv - recommended" ON)
@@ -15,7 +16,7 @@ if(NOT CMAKE_BUILD_TYPE)
         FORCE)
 endif()
 
-find_package(Lua51 REQUIRED)
+find_package(Lua REQUIRED)
 include_directories(${LUA_INCLUDE_DIR})
 
 if(NOT USE_INTERNAL_FPCONV)
@@ -43,6 +44,19 @@ else()
     endif()
 endif()
 
+include(CMakePackageConfigHelpers)
+include(GNUInstallDirs)
+write_basic_package_version_file(
+    "${CMAKE_CURRENT_BINARY_DIR}/cjsonConfigVersion.cmake"
+    VERSION ${PROJECT_VERSION}
+    COMPATIBILITY AnyNewerVersion
+)
+configure_package_config_file(
+    cmake/cjsonConfig.cmake.in
+    ${CMAKE_CURRENT_BINARY_DIR}/cjsonConfig.cmake
+    INSTALL_DESTINATION ${CMAKE_INSTALL_DATADIR}/cjson
+)
+
 # Handle platforms missing isinf() macro (Eg, some Solaris systems).
 include(CheckSymbolExists)
 CHECK_SYMBOL_EXISTS(isinf math.h HAVE_ISINF)
@@ -51,7 +65,7 @@ if(NOT HAVE_ISINF)
 endif()
 
 set(_MODULE_LINK "${CMAKE_THREAD_LIBS_INIT}")
-get_filename_component(_lua_lib_dir ${LUA_LIBRARY} PATH)
+get_filename_component(_lua_lib_dir ${LUA_LIBRARIES} DIRECTORY)
 
 if(APPLE)
     set(CMAKE_SHARED_MODULE_CREATE_C_FLAGS
@@ -60,7 +74,7 @@ endif()
 
 if(WIN32)
     # Win32 modules need to be linked to the Lua library.
-    set(_MODULE_LINK ${LUA_LIBRARY} ${_MODULE_LINK})
+    set(_MODULE_LINK ${LUA_LIBRARIES} ${_MODULE_LINK})
     set(_lua_module_dir "${_lua_lib_dir}")
     # Windows sprintf()/strtod() handle NaN/inf differently. Not supported.
     add_definitions(-DDISABLE_INVALID_NUMBERS)
@@ -71,13 +85,40 @@ endif()
 if(MSVC)
     add_definitions(-D_CRT_SECURE_NO_WARNINGS)
     add_definitions(-Dinline=__inline)
-    add_definitions(-Dsnprintf=_snprintf)
-    add_definitions(-Dstrncasecmp=_strnicmp)
 endif()
 
-add_library(cjson MODULE lua_cjson.c strbuf.c ${FPCONV_SOURCES})
+
+add_library(cjson SHARED lua_cjson.c strbuf.c ${FPCONV_SOURCES})
 set_target_properties(cjson PROPERTIES PREFIX "")
 target_link_libraries(cjson ${_MODULE_LINK})
-install(TARGETS cjson DESTINATION "${_lua_module_dir}")
 
-# vi:ai et sw=4 ts=4:
\ No newline at end of file
+
+if(MSVC)
+    set_target_properties(cjson PROPERTIES WINDOWS_EXPORT_ALL_SYMBOLS ON)
+endif()
+
+export(
+    TARGETS cjson
+    NAMESPACE cjson::
+    FILE ${CMAKE_CURRENT_BINARY_DIR}/cjsonTargets.cmake
+)
+install(
+    FILES ${CMAKE_CURRENT_BINARY_DIR}/cjsonConfig.cmake
+    FILES ${CMAKE_CURRENT_BINARY_DIR}/cjsonConfigVersion.cmake
+    DESTINATION ${CMAKE_INSTALL_DATADIR}/cjson
+)
+install(
+    TARGETS cjson
+    EXPORT cjsonTargets
+    RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
+    LIBRARY DESTINATION ${CMAKE_INSTALL_BINDIR}
+    ARCHIVE DESTINATION ${CMAKE_INSTALL_BINDIR}
+    INCLUDES DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
+)
+install(
+    EXPORT cjsonTargets
+    NAMESPACE cjson::
+    DESTINATION ${CMAKE_INSTALL_DATADIR}/cjson
+)
+
+# vi:ai et sw=4 ts=4:
